include/deal.II/numerics/smoothness_estimator.h

```diff
@@ -0,0 +1,268 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 1998 - 2018 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_smoothness_estimator_h
+#define dealii_smoothness_estimator_h
+
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/fe/fe_series.h>
+
+#include <deal.II/lac/vector.h>
+
+#include <deal.II/numerics/vector_tools.h>
+
+#include <vector>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+
+/**
+ * A namespace for various smoothness estimation strategies for hp-adaptive FEM.
+ */
+namespace SmoothnessEstimator
+{
+  /**
+   * Estimate smoothness from decay of Legendre absolute values of coefficients
+   * on the reference cell.
+   *
+   * In one dimension, the finite element solution on the reference element with
+   * polynomial degree $p$ can be written as
+   * @f[
+   *    u(x) = \sum_{j=0}^{p} a_j P_j(x)
+   * @f]
+   * where $\{P_j(x)\}$ are Legendre polynomials. The decay of the coefficients
+   * is estimated by performing the least squares fit of
+   * @f[
+   *   |a_j| \sim C exp(-\sigma j)
+   * @f]
+   * for $j=1,..,p$. The rate of the decay can be used to estimate the
+   * smoothness. For example, p-refinement will be performed if $\sigma>1$.
+   *
+   * Extension to higher dimension is done by performing the fit in each
+   * direction separately and then taking the lowest value of $\sigma$.
+   *
+   * The returned value for each vector and for each cell is $\exp(-\sigma)$.
```

Comment:
Do you want to say "For each input vector of degrees of freedom defined on a DoFHandler, this function returns a vector with as many elements as there are cells where each element contains $\exp(-\sigma)$."?

Also, why do you return the exponential of sigma, instead of the value itself?

Reply:
> Also, why do you return the exponential of sigma, instead of the value itself?

no particular reason. But in case linear regression fails (i.e. not enough non-zero coefficients in each direction/dimension to fit), this will return `0` (exponent of the default slope minimum infinity). I can also switch it to return `\sigma`.

Reply:
I think `exp(-\sigma)` has no particular meaning. What does have meaning is the function `exp(-\sigma*k)`, but that's an unrelated thing. I'd just return `\sigma`.

Reply:
> I think exp(-\sigma) has no particular meaning.

actually, it does. If you are curious, have a look at Houston2005, eq. 14. I will update the references and extend this sentence.